Add HTTP/3 support

strarsis opened the pull request:
This PR adds the necessary configuration for proper HTTP/3 support (by `nginx`).
Some additional SSL/QUIC/HTTP/3 tweaks are added to the configuration when HTTP/3 support is turned on.
The OS ansible-managed `ferm` firewall is also configured to allow inbound UDP/433 for QUIC (HTTP/3). The approach is also used in the previous PR for HTTPS (https://github.com/roots/trellis/pull/1530).

Additional notes:
- Besides the OS firewall (that is configured by Trellis) there is often a hardware/cloud firewall in front of the server, 
of course it also needs to be configured for allowing inbound UDP/443 traffic for QUIC (HTTP/3).
- HTTP/3 requires an additional HTTP header to be sent for advertising the availability of QUIC/HTTP/3,
without this HTTP header, HTTP/3 will not work, despite everything else working fine.
- Port `443` is not mandatory for HTTP/3, but it is recommended to use the same port `443` as for HTTPS (TCP), `nginx` can listen on QUIC/UDP `443` in parallel to the `443` TCP for HTTPS.
- In my experience, testing HTTP/3 requests worked best with the Chrome Developer Tools (Network tab, toggle `Protocol` column on). 
`curl`/`wget` HTTP/3 support is not a given in current stable Ubuntu. I used a [`curl Docker image with HTTP/3 support`](https://github.com/yurymuski/curl-http3). Some online HTTP/3 testing tools do not always work and incorrectly report HTTP/3 not being supported.

strarsis commented:
@swalkinshaw: Using `nginx` includes makes the configuration much readable now.

strarsis commented:
The SSL early data option allows for RTT-0 requests (zero round-trip time), however, it comes with security implications, the application layer (so the PHP WordPress app here) gets a HTTP Header `Early-Data` passed from the reverse proxy (`nginx`), and for risky operations (as authentication/login) the app has to terminate requests with HTTP `425 Too Early`.
I was not able to find any occurrence of this logic in WordPress code or plugins or discussions about this, hence I commented out this optimization until it is deemed safe for WordPress applications.

swalkinshaw commented:
Mind rebasing @strarsis ? Looks good otherwise and all the notes/documentation is appreciated.

strarsis commented:
@swalkinshaw: Sure! I also have to test the HTTP/3 specific configuration a bit further.

strarsis commented:
@swalkinshaw: Well, I "rebased" it somehow. If necessary, I create a new branch/PR.

So it turned out that a global listen for QUIC with `reuseport` is necessary for working QUIC responses.
For a QUIC listen is also a SSL certificate required, it does not matter which one.
So for providing such a SSL certificate, one of the sites have to be specified as the "default site".

swalkinshaw commented:
😓  wow they really don't make this easy. I'll try and think of another solution for the default site/SSL cert 🤔

strarsis commented:
> another solution for the default site/SSL cert

That listen quic is only needed for reuseport.  With listen quic nginx requires a SSL cert and key, but that listen and SSL would not be used otherwise. Maybe a snakeoil SSL cert would suffice to satisfy the SSL cert condition by nginx?

strarsis commented:
@swalkinshaw: Are Jinja2 namespaces supported in the currently used ansible version (by Trellis CLI init)?`
I need that feature for adding `reuseport` in the configuration of only the first site that uses HTTPS+HTTP/3.
```yml
# {{ ansible_managed }{% set wpsiteconfig = namespace(rendered_first_quic=false) %}

{% block server_before %}{% endblock %}

server {
  {% block server_id -%}
  listen {{ ssl_enabled | ternary('[::]:443 ssl', '[::]:80') }};
  listen {{ ssl_enabled | ternary('443 ssl', '80') }};

  {% if nginx_http3_enabled and ssl_enabled -%}
  # Listen on UDP for QUIC+HTTP/3 '{{ wpsite.rendered_first_quic }}'
  listen [::]:443 quic{% if not wpsiteconfig.rendered_first_quic %} reuseport{% endif -%};
  listen 443 quic{% if not wpsiteconfig.rendered_first_quic %} reuseport{% endif -%};
  {% if not wpsiteconfig.rendered_first_quic -%}
  # there has to be one listen quic directive with reusport for working QUIC responses in current nginx version, using the first site.
  {% set wpsiteconfig.rendered_first_quic = "true" %}{% endif -%}
  {% endif -%}

  http2 {{ nginx_http2_enabled | default(false) | ternary('on', 'off') }};
  http3 {{ nginx_http3_enabled | default(false) | ternary('on', 'off') }};
  server_name {{ site_hosts_canonical | union(multisite_subdomains_wildcards) | join(' ') }};
  {% endblock %}
[...]
````
The first line appears commented out, but that is a `nginx` configuration comment (`#`), not a jina2 one btrw.

strarsis commented:
@swalkinshaw: Edit: After some real-world testing I noticed that some WordPress sites had redirect issues (on frontend) (`ERR_TOO_MANY_REDIRECTS`). I am not sure whether this has been caused by http/3. Still testing and observing the behavior with http/3 enabled. [Possible related issue](https://mailman.nginx.org/pipermail/nginx/2021-July/060879.html).

This appears to be caused by the redirects, when HTTP/3 is advertised/used there.

swalkinshaw commented:
Plug might make sense, but at least for now we can just implement the 425 for non-GET requests in Nginx right? To keep it simple.

I think once we do that the PR should be ready to merge

swalkinshaw commented on roles/wordpress-setup/templates/includes/directive-only/http3-rtt0.conf:
```diff
@@ -0,0 +1,8 @@
+proxy_set_header Early-Data $ssl_early_data;
+
+# HTTP/3 Early-Data/0-RTT replay attack mitigration for non-GET requests
+if ($ssl_early_data = 1 && $request_method !~ ^(GET|HEAD|OPTIONS)$) {
```
Nginx doesn't support multiple conditions with `&&` I don't think? We'd need a `map` like this afaik:

```nginx
map "$ssl_early_data:$request_method" $reject_early_data {
    default 0;
    "~^1:(?!GET|HEAD|OPTIONS)" 1;
}

# In server block:
if ($reject_early_data) {
    return 425;
}
```

strarsis commented:
@swalkinshaw: Thanks for the review, I overhauled some of the PR components, testing it locally before pushing.
Incredibly busy right now, next week I should have more time getting this into an merge-acceptable state.

strarsis commented on roles/wordpress-setup/templates/includes/directive-only/http3-negotiate-redirect.conf:
```diff
@@ -0,0 +1,2 @@
+# Add Alt-Svc header to negotiate HTTP/3 (when redirecting from HTTP).
+add_header alt-svc 'h3-27=":443"; ma=86400, h3-28=":443"; ma=86400, h3-29=":443"; ma=86400, h3=":443"; ma=86400';
```
Yes, this is not needed, I consolidated this into one single configuration file.

strarsis commented:
I need to spin up some VPS and test the whole thing (HTTP/3; QUIC; RTT-0 (when enabled)) with this PR.

retlehs commented:
@strarsis I had to make these changes to provision successfully with this PR:

```diff
diff --git a/trellis/roles/wordpress-setup/defaults/main.yml b/trellis/roles/wordpress-setup/defaults/main.yml
index 4eff909..408e627 100644
--- a/trellis/roles/wordpress-setup/defaults/main.yml
+++ b/trellis/roles/wordpress-setup/defaults/main.yml
@@ -1,11 +1,12 @@
 site_uses_local_db: "{{ site_env.db_host == 'localhost' }}"
+wordpress_default_site: "{{ wordpress_sites.keys() | first }}"
 nginx_wordpress_site_conf: wordpress-site.conf.j2
 nginx_ssl_path: "{{ nginx_path }}/ssl"
 
 nginx_sites_confs:
   - src: no-default.conf.j2
   - src: http3-reuseport.conf.j2
-    enabled: "{{ nginx_http3_enabled and (sites_use_ssl | bool) and wordpress_default_site }}"
+    enabled: "{{ nginx_http3_enabled and (sites_use_ssl | bool) and (wordpress_default_site | length > 0) }}"
   - src: ssl.no-default.conf.j2
     enabled: false
 
diff --git a/trellis/roles/wordpress-setup/templates/wordpress-site.conf.j2 b/trellis/roles/wordpress-setup/templates/wordpress-site.conf.j2
index 5e0d2ea..b3b7772 100644
--- a/trellis/roles/wordpress-setup/templates/wordpress-site.conf.j2
+++ b/trellis/roles/wordpress-setup/templates/wordpress-site.conf.j2
@@ -9,10 +9,8 @@ server {
 
   {% if nginx_http3_enabled and ssl_enabled -%}
   # Listen on UDP for QUIC+HTTP/3
-  listen [::]:443 quic{% if is_first_site_use_ssl %} reuseport{% endif -%};
-  listen 443 quic{% if is_first_site_use_ssl %} reuseport{% endif -%};
-  {% if is_first_site_use_ssl -%}# there has to be one listen quic directive with `reuseport` for working QUIC responses in current nginx version, using the first site.
-  {% endif %}
+  listen [::]:443 quic;
+  listen 443 quic;
   {% endif %}
 
   http2 {{ nginx_http2_enabled | default(false) | ternary('on', 'off') }};
```

Also should be worth noting at the top of this PR that it must be enabled by updating `trellis/roles/wordpress-setup/defaults/main.yml`:

```diff
- nginx_http3_enabled: false
+ nginx_http3_enabled: true
```

strarsis commented:
@swalkinshaw, @retlehs: Yes, I agree and removed the 0-RTT config from this PR now.

0-RTT support is highly experimental in nginx and also not directly related to HTTP, it is a TLS1.3 optimization feature with HTTP header support on top ([Early data](https://developer.mozilla.org/de/docs/Web/HTTP/Reference/Headers/Early-Data)). Client support is also lacking: [Only Firefox currently supports it as in March 2026](https://caniuse.com/wf-early-data).
So when 0-RTT finally becomes widespread in client and server support, we can revisit this in a new PR.
Honestly I do not even know why I have added 0-RTT to this PR in the first place, probably because HTTP/3 had been more experimental when I started the PR and 0-RTT was equally new then, and often discussed as a bleeding-edge optimization technique.
